### DISCUSSION TITLE
how to publish a message when client disconnects

### DISCUSSION THREAD

**Discussion by mangroovie** [empicano/aiomqtt#91] [Q&A] [answered]
I have service which should report its state in a topic /service/state as online/offline
For normal operation I can publish state online to /service/state. when service starts as shown in examples.
For unexpected disconnects I can use will to set the /service/state to offline and broker should handle these ok.

But how about normal disconnect? when service is stopped or Keyboard interrupted? 
How to ensure the state is published before client disconnects?
In vanilla paho.mqtt.client this would be handled on_disconnect callback but how to do this with asyncio_mqtt?

**Comment by frederikaalund** [comment, marked as answer]
> For unexpected disconnects I can use will to set the /service/state to offline and broker should handle these ok.

That's a good use of the `will` feature in MQTT. :+1: 

> But how about normal disconnect? when service is stopped or Keyboard interrupted?
How to ensure the state is published before client disconnects?

Good question. :) Sounds like a case for the `finally` block. I suggest something like the following (adapted from the basic example in `README.md`):

```python
async with Client("test.mosquitto.org") as client:
    try:
        async with client.filtered_messages("floors/+/humidity") as messages:
            await client.subscribe("floors/#")
            async for message in messages:
                print(message.payload.decode())
    # The `finally` block always runs (both on exception and otherwise)
    finally:
        await client.publish("service/state", payload="offline".encode())
```

Does it make sense?

**Reply by mangroovie** [reply]
I'm actually using the advanced example and AsyncExitStack to create new top tasks for each topic and as in the example tasks are gathered in the end

**Reply by mangroovie** [reply]
Got it working by adapting the shown snippet. Thanks!

**Comment by frederikaalund** [comment]
On an unrelated note: Avoid a leading forward slash in MQTT topics. That is, prefer `service/state` over `/service/state`.

Reference: https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/

**Reply by mangroovie** [reply]
Sure. That is a subpath of topic I copied so no leading slashes in reality.
